python/tests/test_topology.py

```diff
@@ -4984,3 +4984,62 @@ def test_ts_single_tree_delete_middle(self):
         ts_keep = ts.delete_intervals([[0.25, 0.5]], record_provenance=False)
         ts_delete = ts.keep_intervals([[0, 0.25], [0.5, 1.0]], record_provenance=False)
         self.assertTrue(ts_equal(ts_keep, ts_delete))
+
+
+class TestTrim(unittest.TestCase):
+    """
+    Test the trimming functionality
+    """
```

Comment:
This is a good start. You need to add some tests which include sites as well (not testing at all at the moment, right?)

The tests aren't very thorough either --- I could implement trim at the moment by just putting in a single edge from left to right, and I think the tests would pass. Need to test that the bits in the middle have been faithfully kept (I guess, comparing the trees is the easiest? parent_dict should be equal for each pair of trees in the before and after ts).

The coordinates are all very nice and clean right now too. I'm worried about the case where we have a value like 0.1 which doesn't represent exactly in binary. Should we force the left coordinate to be exactly zero, or do we accept the fact that there'll be a tiny gap at the start of the sequence? Or worse, could subtracting the left coordinate lead to a negative number?

There's all sorts of corner cases that need to be explored here.

Reply:
> This is a good start. You need to add some tests which include sites as well (not testing at all at the moment, right?)

Good point. I've just added mutations in. Sorry for the oversight.

> The tests aren't very thorough either --- I could implement trim at the moment by just putting in a single edge from left to right, and I think the tests would pass. Need to test that the bits in the middle have been faithfully kept (I guess, comparing the trees is the easiest? parent_dict should be equal for each pair of trees in the before and after ts).

Done in latest push

> The coordinates are all very nice and clean right now too. I'm worried about the case where we have a value like 0.1 which doesn't represent exactly in binary. Should we force the left coordinate to be exactly zero, or do we accept the fact that there'll be a tiny gap at the start of the sequence? Or worse, could subtracting the left coordinate lead to a negative number?

Where would the negative number come from? If we subtract the left of the leftmost edge, won't it simply make all the other coords not-quite-simple numbers?

Would it be enough to test with (say) 0.1 rather than 0.25 as I do at present?

Reply:
I don't know --- odd things happen at the boundaries when you have very small numbers. Try out a left coordinate of DBL_MIN or someting.

Rather than jamming everything into a handful of tests, write some more tests. Try out different corner cases in different named tests --- that way, a reader of the code can see what that specific test is for.

A 'verify' function here which takes a generic tree sequence as input and checks the output seems like a good idea. Simple specific tests are good, but there needs to be a more general checks to you can verify a bunch of different examples as well.

For example, you've not tested sites that are exactly on the boundaries here.

Reply:
Yes, you are right, I should separate some of this stuff into separate tests.